File: Problem Questions/Week 01: Arrays & Strings/Easy/Find max & min element in array/solution.py
''' 
# Find Maximum and Minimum in Array

# Problem Statement

# Given an array of integers, return the **maximum** and **minimum** elements in the array.

# Example


**Input:**  
`nums = [3, 1, 5, 9, 0, -3]`

**Output:**  
`max = 9, min = -3`
'''

from typing import List


def max_and_min(array: List[int]) -> str:
    # set the index of the first value of the array (0) to both the max and the min.
    # loop through the array once updating values of max and min
    # when a smaller or larger value is found

    if not array:  # edge case (empty array)
        return "Array is empty."

    max = 0
    min = 0
    size = len(array)

    for i in range(1, size):
        if array[i] < array[min]:
            min = i
        if array[i] > array[max]:
            max = i

    return f"max = {array[max]}, min = {array[min]}"


# Example Test Cases

nums = [3, 1, 5, 9, 0, -3]

nums2 = [3, 1, 5, 9, 0, -3, 6, 12, 4, 687, 654, 312, -4, -15, -213]

nums3 = [5]  # in the case of a single element in the array

nums4 = []  # in the case of an empty array

print(max_and_min(nums))
print(max_and_min(nums2))
print(max_and_min(nums3))
print(max_and_min(nums4))


# Time Complexity : O(n)
# Space Complexity: O(1)

File: Problem Questions/Week 01: Arrays & Strings/Easy/Find max & min element in array/C/solution.c
/*
# Find Maximum and Minimum in Array

## Problem Statement

Given an array of integers,
return the **maximum** and **minimum**
elements in the array.

---

## Example

**Input:**
`nums = [3, 1, 5, 9, 0, -3]`

**Output:**
`max = 9, min = -3`

---
*/

#include <stdio.h>

// function prototype:
void findMaxMin(int *array, int length, int *min, int *max);

int main()
{
    // usage
    int nums[] = {21, 78, 32, 4, 6, 23, 3, 12, 1, 2};
    int min, max;
    int length = sizeof(nums) / sizeof(nums[0]);

    findMaxMin(nums, length, &min, &max);

    printf("Min: %d, Max: %d\n", min, max);

    return 0;
}

// Approach: Grab a page off selection sort logic
void findMaxMin(int *array, int length, int *min, int *max)
{
    // Step 1: Initialize a max and min variable at index 0
    *min = array[0];
    *max = array[0];

    // Step 2: Loop through array updating the values,
    // when a new max or min is found.
    for (int i = 0; i < length; i++)
    {
        if (array[i] < *min)
        {
            *min = array[i];
        }
        if (array[i] > *max)
        {
            *max = array[i];
        }
    }
}
